object_intern.go

```diff
@@ -580,9 +587,71 @@ func (oi *ObjectIntern) DeleteBatch(ptrs []uintptr) {
 			}
 
 			// most likely case is that we will just decrement the reference count and return
-			if atomic.LoadUint32((*uint32)(unsafe.Pointer(p+uintptr(len(obj)-4)))) > 1 {
+			if atomic.LoadUint32((*uint32)(unsafe.Pointer(p))) > 1 {
+				// decrement reference count by 1
+				atomic.AddUint32((*uint32)(unsafe.Pointer(p)), ^uint32(0))
+				continue
+			}
+
+			// if reference count is 1 or less, delete the object and remove it from index
+			// If one of these operations fails it is still safe to perform the other
+			// Once we get to this point we are just going to remove all traces of the object
+
+			// delete object from index first
+			// If you delete all of the objects in the slab then the slab will be deleted
+			// When this happens the memory that the slab was using is MUnmapped, which is
+			// the same memory pointed to by the key stored in the ObjIndex. When you try to
+			// access the key to delete it from the ObjIndex you will get a SEGFAULT
+			//
+			// remove 4 leading bytes for reference count since ObjIndex does not store reference count in the key
+			delete(oi.objIndex, string(obj[4:]))
+
+			// delete object from object store
+			err = oi.store.Delete(p)
+		}
+
+		oi.Unlock()
+	}
+}
+
+// DeleteBatchUnsafe does the same thing as DeleteBatch, but saves time by not acquiring
+// read locks if the objects only need their reference count decremented. This is not safe, and it
+// is up to the caller to ensure the objects actually exist in the store. If you are unsure, don't use this
+// method.
+func (oi *ObjectIntern) DeleteBatchUnsafe(ptrs []uintptr) {
+
+	toDelete := ptrs[:0]
```

Comment:
this would be a smaller allocation if toDelete was make([]bool,len(ptrs)) to keep track of which need to be deleted